### DISCUSSION TITLE
ble_uart_repl.py

### DISCUSSION THREAD

**Discussion by patfrench** [micropython/micropython#12661] [General]
Hello,
I don't understand how the script works: ble_uart_repl.py

In my GOOGLE console I get this when I type 3+3 in REPL 

![image](https://github.com/micropython/micropython/assets/44723412/a4794f8a-6e03-4eb7-860f-4ca1e8849b1c)

I think I don't understand how it works

Could someone explain it to me?

Thank you

**Comment by andrewleech** [comment]
Could you give a few more details please:
* Where did this ble_uart_repl.py script come from?
* what board are you running it on?
* what do you mean by Google console?
* how are you running the script? 
I'm guessing thonny by the screenshot, I think thonny does some weird stuff to the console, it might be breaking the repl.

**Comment by patfrench** [comment]
Thank you !

- https://github.com/micropython/micropython/blob/master/examples/bluetooth/ble_uart_repl.py
- ESP32
- I developed a javascript program with the webbluetooth API. 
I detect my ESP32
send data to esp32 with writecharacteristic
but when I write in THONNY's REPL then I receive data in my google console

- I just added a start() at the end of the script

I don't quite understand what the python script does: I'd like to send data in str format 
javascript send => "print("test")"
ESP32 receives and executes as eval() would

Thank you !

**Reply by andrewleech** [reply]
Inside micropython, there is only one repl instance that's communicated to via stdin / stdout. By default at boot, stdin/out are connected to the units usb and/or uart - on some boards that have both uart and usb you can connect both to separate terminal windows and see everything typed in one show in the other etc.

That's basically what will be happening here; this script uses [dupterm](https://docs.micropython.org/en/latest/library/os.html#terminal-redirection-and-duplication) to add another stream of I/O to the same internal stdio / repl interface. This means that anything typed into one interface will show on the other as well.

Thonny might complicate this as it sends other control-character things over the repl interface that might show up really strangely on a ble interface, I'm not sure though as I've never used thonny myself.

**Reply by patfrench** [reply]
Hi,
Thank you for your explain.

it works very well with putty.
I have a question : when javascript send print("test"), the REPL receives print("test"), but how can I make it execute automatically ?

I tried "Raw mode and raw-paste mode" (with CTRL+A) but it doesn't work

**Reply by andrewleech** [reply]
Did you try adding `\n` or `\r\n` on the end of the string from javascript?

**Reply by patfrench** [reply]
I tried but the REPL receives "print("test")\n\r".

I think the problem is with my JavaScript. JavaScript sends str when the UART wants bytes.

I'm going to modify my JavaScript
